### ISSUE TITLE
New libc++ test configs do not pass sufficient envvars to .sh.cpp tests

### ISSUE THREAD

**Issue by mgorny** [llvm/llvm-project#56816]
I've been investigating the following test failure with the 15.x branch:

```console
FAIL: llvm-libc++abi-shared.cfg.in :: incomplete_type.sh.cpp (29 of 71)
******************** TEST 'llvm-libc++abi-shared.cfg.in :: incomplete_type.sh.cpp' FAILED ********************
[...]
Command Output (stdout):                                                                                                               
--                                                                                                                                     
$ ":" "RUN: at line 22"                                                                                                                
$ "/usr/lib/ccache/bin/x86_64-pc-linux-gnu-clang++" "--target=x86_64-pc-linux-gnu" "-nostdinc++" "-I" "/tmp/portage/sys-libs/libcxxabi-15.0.0.9999/work/libcxxabi/include" "-I" "/tmp/portage/sys-libs/libcxxabi-15.0.0.9999/work/runtimes_build-abi_x86_64.amd64/include/c++/v1" "-I" "/tmp/portage/sys-libs/libcxxabi-15.0.0.9999/work/runtimes_build-abi_x86_64.amd64/include/x86_64-pc-linux-gnu/c++/v1" "-I" "/tmp/portage/sys-libs/libcxxabi-15.0.0.9999/work/libcxxabi/../libcxx/test/support" "-I" "/tmp/portage/sys-libs/libcxxabi-15.0.0.9999/work/libcxxabi/../libcxx/src" "-D_LIBCPP_ENABLE_CXX17_REMOVED_UNEXPECTED_FUNCTIONS" "-std=c++2b" "-Werror" "-Wall" "-Wextra" "-Wshadow" "-Wundef" "-Wno-unused-command-line-argument" "-Wno-attributes" "-Wno-pessimizing-move" "-Wno-c++11-extensions" "-Wno-noexcept-type" "-Wno-atomic-alignment" "-Wno-user-defined-literals" "-Wno-tautological-compare" "-Wsign-compare" "-Wunused-variable" "-Wunused-parameter" "-Wunreachable-code" "-Wno-unused-local-typedef" "-D_LIBCPP_HAS_NO_PRAGMA_SYSTEM_HEADER" "-D_LIBCPP_ENABLE_ASSERTIONS=1" "-D_LIBCPP_DISABLE_AVAILABILITY" "-fcoroutines-ts" "-Werror=thread-safety" "-Wuser-defined-warnings" "-Wno-unreachable-code" "-c" "/tmp/portage/sys-libs/libcxxabi-15.0.0.9999/work/libcxxabi/test/incomplete_type.sh.cpp" "-o" "/tmp/portage/sys-libs/libcxxabi-15.0.0.9999/work/runtimes_build-abi_x86_64.amd64/Output/incomplete_type.sh.cpp.dir/t.tmp.one.o"
# command stderr:
ccache: error: Failed to create temporary file for /var/lib/portage/home/.cache/ccache/tmp/tmp.cpp_stdout.XXXXXX: Permission denied

error: command failed with exit status: 1
```

It seems that the problem is that the new test configs (`llvm-libc++abi-shared.cfg.in` here) do not set `config.environment` like the legacy config did. As a result, lit only passes a small subset of envvars to the test commands. In this particular case, it strips `HOME`, `XDG_CACHE_HOME` and `CCACHE_DIR`, and ccache tries to write into the wrong directory as a result. If I add a quick debug, at the point of configuration, `config.environment` is stripped down to:

```python
{'PATH': '/tmp/portage/sys-libs/libcxxabi-15.0.0.9999/temp/python3.10/bin:/usr/lib/ccache/bin:/usr/lib/portage/python3.11/ebuild-helpers:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/opt/bin:/usr/lib/llvm/15/bin:/usr/lib/llvm/15/bin:/usr/lib/llvm/14/bin', 'LLVM_DISABLE_CRASH_REPORT': '1', 'TERM': 'tmux-256color', 'LD_PRELOAD': 'libsandbox.so', 'TMPDIR': '/tmp/portage/sys-libs/libcxxabi-15.0.0.9999/temp', 'TMP': '/tmp/portage/sys-libs/libcxxabi-15.0.0.9999/temp', 'TEMP': '/tmp/portage/sys-libs/libcxxabi-15.0.0.9999/temp'}
```

I suspect libc++ may be suffering from the same problem but I haven't gotten that far yet.

CC @ldionne

**Comment by mgorny** [comment]
Yep, libc++ suffers from the same problem.

**Comment by mgorny** [comment]
libunwind also suffers from the same problem, though the symptoms there are even more cryptic:

```pytb
lit: /usr/lib/python3.10/site-packages/lit/TestingConfig.py:136: fatal: unable to parse config file '/tmp/portage/sys-libs/llvm-libunwi
nd-15.0.0.9999/work/runtimes_build-abi_x86_32.x86/libunwind/test/lit.site.cfg', traceback: Traceback (most recent call last):
  File "/usr/lib/python3.10/site-packages/lit/TestingConfig.py", line 125, in load_from_path
    exec(compile(data, path, 'exec'), cfg_globals, None)
  File "/tmp/portage/sys-libs/llvm-libunwind-15.0.0.9999/work/runtimes_build-abi_x86_32.x86/libunwind/test/lit.site.cfg", line 36, in <
module>
    libcxx.test.newconfig.configure(
  File "/tmp/portage/sys-libs/llvm-libunwind-15.0.0.9999/work/libcxx/utils/libcxx/test/newconfig.py", line 22, in configure
    actions = param.getActions(config, lit_config.params)
  File "/tmp/portage/sys-libs/llvm-libunwind-15.0.0.9999/work/libcxx/utils/libcxx/test/dsl.py", line 712, in getActions
    (_, parameterValue) = self._getValue(config, litParams)
  File "/tmp/portage/sys-libs/llvm-libunwind-15.0.0.9999/work/libcxx/utils/libcxx/test/dsl.py", line 691, in _getValue
    value = getDefault()
  File "/tmp/portage/sys-libs/llvm-libunwind-15.0.0.9999/work/libcxx/utils/libcxx/test/dsl.py", line 686, in <lambda>
    getDefault = lambda: self._default(config) if callable(self._default) else self._default
  File "/tmp/portage/sys-libs/llvm-libunwind-15.0.0.9999/work/libcxx/utils/libcxx/test/params.py", line 80, in <lambda>
    default=lambda cfg: next(s for s in reversed(_allStandards) if getStdFlag(cfg, s)),
StopIteration

```

**Comment by mgorny** [comment]
Submitted https://reviews.llvm.org/D130843 as a proposed fix.

**Comment by mgorny** [comment]
@ldionne, could you confirm that it's ok to backport to 15.x?

**Comment by ldionne** [comment]
Yes, it is, but please let me do it. I'm handling backports to libc++ myself in bulk.

**Comment by tstellar** [comment]
Merged: b0eab153405eaa29653f86ff13fa1832e0fbf991

**Comment by mgorny** [comment]
Thanks!
